### PR TITLE
Loosen some argument types

### PR DESCRIPTION
**Pull request by Xiretza**
This supersedes #13 and additionally removes some unnecessary cloning.

### CODE REVIEW

File: src/service.rs [hunk 1]
```diff
@@ -612,23 +612,20 @@ impl CollectionManager {
...
     /// * `options` - parameters to tune or optimize the fetch
-    pub fn list_multi<'a, I>(
+    pub fn list_multi<I>(
         &self,
         collection_types: I,
         options: Option<&FetchOptions>,
     ) -> Result<CollectionListResponse<Collection>>
     where
-        I: Iterator<Item = &'a str>,
+        I: IntoIterator,
```
Comment (tasn) [inline src/service.rs:621]: Correct me if I'm wrong, but this forces a clone at times, while the former is just a reference with a limited lifetime? Would this change be potentially worse?

File: src/online_managers.rs [hunk 2]
```diff
@@ -493,23 +493,23 @@ impl CollectionManagerOnline {
...
 
-    pub fn list_multi<'a, I>(
+    pub fn list_multi<I>(
         &self,
         collection_types: I,
         options: Option<&FetchOptions>,
     ) -> Result<CollectionListResponse<EncryptedCollection>>
     where
-        I: Iterator<Item = &'a [u8]>,
+        I: IntoIterator<Item = Vec<u8>>,
```
Comment (tasn) [inline src/online_managers.rs:502]: Same as the other lifetime comment.

File: src/service.rs (hunk 1 above)
Comment (Xiretza) [reply src/service.rs:621]: No, this change simply generalizes the `&str` item type to also allow passing `String` and such. The difference between this and the `ItemMetadata::set_*` functions is that we don't actually need a `String` in this function body, only a `&str` (to pass to `collection_type_to_uid`), so taking an argument that's "something like a `&str`" is fine.

File: src/online_managers.rs (hunk 2 above)
Comment (Xiretza) [reply src/online_managers.rs:502]: This change is a little trickier to justify. True, it used to work with just passing in borrowed byte slices and now requires owned `Vec`s, so if a caller only had the former they'd have to do clones. However, this function is internal and not public, so we know all its callers - in fact there's only one, `CollectionManager::list_multi`. This caller already has a `Vec<u8>` (well, `String`), so there's no need to allow passing in a slice; in fact requiring a slice to passed in would make it impossible to remove the `.collect()` in `CollectionManager::list_multi`.

File: src/service.rs (hunk 1 above)
Comment (tasn) [reply src/service.rs:621]: I'm talking about `IntoIterator` doesn't that take ownership?

File: src/online_managers.rs (hunk 2 above)
Comment (tasn) [reply src/online_managers.rs:502]: OK, I'm OK with this if it's internal.

File: src/service.rs (hunk 1 above)
Comment (Xiretza) [reply src/service.rs:621]: Every `Iterator<Item = I>` is also trivially `IntoIterator<Item = I>`, so for any type that was accepted previously, there is no change. This additionally allows directly passing collections without having to call `.into_iter()`/`.iter()` on them manually.

Comment (tasn) [reply src/service.rs:621]: Is it though? I though they are the same the other way around (so every IntoIterator is also Iterator). Anyhow, OK, not a biggie.

Comment (Xiretza) [reply src/service.rs:621]: No, see this impl: https://doc.rust-lang.org/core/iter/trait.IntoIterator.html#impl-IntoIterator-8

`Iterator` is the main trait behind rust iteration, `IntoIterator` is a trait that allows creating an `Iterator` from a collection (e.g. `Vec`), which is a potentially not-entirely-trivial operation. `IntoIterator` is also implemented for all `Iterator`s, but in that case, it's simply an identity function: https://doc.rust-lang.org/src/core/iter/traits/collect.rs.html#263-265

Comment (tasn) [reply src/service.rs:621]: Ah OK then.

**Pull request merged**